### PR TITLE
Initialize init tree forward

### PR DESCRIPTION
**Pull request by pongad**
Instead of creating lists then merging them, we create and populate only one tree. The current implementation creates a lot of temporary objects. I decided to leave it be for now because the PR is already quite large. We can fix this up later if needed.

### CODE REVIEW

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java [hunk 1]
```diff
@@ -70,47 +71,62 @@ public void setupClass() {
   public void testSimpleField() throws Exception {
     String fieldSpec = "myfield";
 
-    InitCodeNode expectedStructure = InitCodeNode.create("myfield");
+    InitCodeNode expectedStructure = InitCodeNode.newRoot();
+    expectedStructure.mergeChild(InitCodeNode.create("myfield"));
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: s/`"myfield"`/`fieldSpec`/

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java [hunk 2]
```diff
@@ -120,58 +136,55 @@ public void testNestedListField() throws Exception {
...
   }
 
   @Test
   public void testFormattedField() throws Exception {
     String fieldSpec = "name%entity";
 
-    HashMap<String, InitValueConfig> initValueMap = new HashMap<>();
     InitValueConfig initValueConfig = InitValueConfig.create("test-api", null);
-    initValueMap.put("name", initValueConfig);
+    InitCodeNode outerStructure = InitCodeNode.createWithValue("name", initValueConfig);
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: "entity" does not show up anywhere because it has no `=`? Could we add comments to that effect here and in the code?

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java [hunk 3]
```diff
@@ -120,58 +136,55 @@ public void testNestedListField() throws Exception {
...
-    Truth.assertThat(
-            actualFormattedFieldNode.getInitValueConfig().hasFormattingConfigInitialValues())
-        .isTrue();
-    Truth.assertThat(
-            actualFormattedFieldNode
-                .getInitValueConfig()
-                .getResourceNameBindingValues()
-                .equals(expectedCollectionValues))
-        .isTrue();
+    assertThat(actualFormattedFieldNode.getInitValueConfig().getResourceNameBindingValues())
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: The previous tests all set up `expectedStructure` and compared it to `FieldStructureParser`. Here we are abandoning the pattern and instead are testing `InitCodeNdoe.createTree()`. Why the change? Could we do this consistently?

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java [hunk 4]
```diff
@@ -221,24 +249,29 @@ public void testEmbeddedFieldList() throws Exception {
...
+    FieldStructureParser.parse(actualStructure, fieldSpec);
+
+    assertNodeEqual(actualStructure, expectedStructure);
   }
 
   @Test(expected = IllegalArgumentException.class)
   public void testFormattedFieldBadField() throws Exception {
     String fieldSpec = "name%entity";
-    FieldStructureParser.parse(fieldSpec);
+    FieldStructureParser.parse(InitCodeNode.newRoot(), fieldSpec);
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: Could we add a comment to this and the following saying the error is because no `InitValueConfig` containing `name` was provided?

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java [hunk 5]
```diff
@@ -120,58 +136,55 @@ public void testNestedListField() throws Exception {
...
-
-    HashMap<String, InitValue> expectedCollectionValues = new HashMap<>();
-    expectedCollectionValues.put("entity1", InitValue.createLiteral("test1"));
-    expectedCollectionValues.put("entity2", InitValue.createLiteral("test2"));
 
     InitCodeContext context =
         getContextBuilder()
             .initFieldConfigStrings(fieldSpecs)
-            .initValueConfigMap(initValueMap)
+            .initValueConfigMap(ImmutableMap.of("formatted_field", initValueConfig))
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: So we only need to provide the formatted field `formatted_field`, and not the components `entity1`, `entity2` that go into it? I should sit down to understand better how this works. We should document it.

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java [hunk 6]
```diff
@@ -350,92 +369,62 @@ public void testMultipleFormattedEntries() throws Exception {
...
             .build();
     InitCodeNode rootNode = InitCodeNode.createTree(context);
-    List<String> actualKeyList = new ArrayList<>();
-    for (InitCodeNode node : rootNode.listInInitializationOrder()) {
-      actualKeyList.add(node.getKey());
-    }
-    Truth.assertThat(actualKeyList.equals(expectedKeyList)).isTrue();
+    assertThat(rootNode.listInInitializationOrder().stream().map(node -> node.getKey()))
+        .containsExactly("formatted_field", "root")
+        .inOrder();
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: Can we test `entity1`, `entity2`?

File: src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java [hunk 7]
```diff
@@ -28,13 +28,15 @@
 public class FieldStructureParser {
   private static final String PROJECT_ID_TOKEN = "$PROJECT_ID";
 
-  public static InitCodeNode parse(String initFieldConfigString) {
-    return parse(initFieldConfigString, ImmutableMap.<String, InitValueConfig>of());
+  public static void parse(InitCodeNode root, String initFieldConfigString) {
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java]: Done

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 8]
```diff
@@ -196,60 +223,45 @@ private InitCodeNode(
...
-        }
-      }
-    }
-
-    for (String initFieldConfigString : context.initFieldConfigStrings()) {
-      subTrees.add(FieldStructureParser.parse(initFieldConfigString, context.initValueConfigMap()));
-    }
-    subTrees.addAll(context.additionalInitCodeNodes());
-    return subTrees;
+  private static <T> T firstNotNull(T a, T b) {
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: Derp. Done

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 9]
```diff
@@ -157,12 +157,39 @@ public static InitCodeNode createSingletonList(String key) {
...
+        context.initFields() != null || context.outputType() != InitCodeOutputType.FieldList,
+        "init field array is not set for flattened method");
+    InitCodeNode root = newRoot();
+
+    if (context.initFields() != null) {
+      for (FieldModel field : context.initFields()) {
+        String nameString = field.getNameAsParameter();
+        InitValueConfig initValueConfig = context.initValueConfigMap().get(nameString);
+        if (initValueConfig == null) {
+          root.mergeChild(InitCodeNode.createWithName(nameString, nameString));
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: It will happen recursively, but the new node doesn't have children, so it won't actually recur.

We still need `mergeChild` for `additionalInitCodeNodes` which might actually be recursive though.

File: src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java [hunk 10]
```diff
@@ -177,12 +130,60 @@ private static InitCodeNode parseConfig(
...
     if (valueConfig != null) {
-      parent.updateInitValueConfig(valueConfig);
+      parent.updateInitValueConfig(
+          InitCodeNode.mergeInitValueConfig(parent.getInitValueConfig(), valueConfig));
       parent.setLineType(InitCodeLineType.SimpleInitLine);
     }
     return root;
   }
 
+  private static InitCodeNode parsePath(InitCodeNode root, Scanner scanner) {
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java]: Done

File: src/main/java/com/google/api/codegen/util/Scanner.java [hunk 11]
```diff
@@ -182,6 +183,10 @@ public int pos() {
     return loc;
   }
 
+  public int last() {
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/util/Scanner.java]: Sounds reasonable.

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 12]
```diff
@@ -157,12 +157,39 @@ public static InitCodeNode createSingletonList(String key) {
     return InitCodeNode.createWithChildren(key, InitCodeLineType.ListInitLine, child);
   }
 
+  public static InitCodeNode newRoot() {
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: Done

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 13]
```diff
@@ -196,60 +223,45 @@ private InitCodeNode(
...
+      oldChild.mergeChild(newSubChild);
     }
+    return oldChild;
   }
 
-  private static InitValueConfig mergeInitValueConfig(
+  static InitValueConfig mergeInitValueConfig(
       InitValueConfig oldConfig, InitValueConfig newConfig) {
-    HashMap<String, InitValue> collectionValues = new HashMap<>();
     if (oldConfig.hasSimpleInitialValue()
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: `Objects.equals` is better :)

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 14]
```diff
@@ -196,60 +223,45 @@ private InitCodeNode(
...
-      collectionValues.putAll(newConfig.getResourceNameBindingValues());
-    }
-    return oldConfig.withInitialCollectionValues(collectionValues);
+    return InitValueConfig.newBuilder()
+        .setApiWrapperName(
+            firstNotNull(oldConfig.getApiWrapperName(), newConfig.getApiWrapperName()))
+        .setSingleResourceNameConfig(
+            firstNotNull(
+                oldConfig.getSingleResourceNameConfig(), newConfig.getSingleResourceNameConfig()))
+        .setInitialValue(firstNotNull(oldConfig.getInitialValue(), newConfig.getInitialValue()))
```
[the remark this answers is not in the document]
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: Done

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 2 above)
Comment (pongad) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: I honestly don't know why this is happen actually, but the test passing implies we're correctly keeping the old behavior.

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 3 above)
Comment (pongad) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: Previously `FieldStructureParser` parses things into a list. Now we have it update the existing tree. So we now have to compare against a tree than a list.

We are doing this consistently right?

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 4 above)
Comment (pongad) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: Done

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 5 above)
Comment (pongad) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: I don't understand this either. I don't think this should block us, but I should also investigate.

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 6 above)
Comment (pongad) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: I don't think we can. The `entity1` and `entity2` will be missing from the data structure for the same reason as [here](https://github.com/googleapis/gapic-generator/pull/2178#discussion_r208031764). I don't really know what the reason is.

File: src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java [hunk 15]
```diff
@@ -177,18 +132,70 @@ private static InitCodeNode parseConfig(
...
+      parent.updateInitValueConfig(
+          InitCodeNode.mergeInitValueConfig(parent.getInitValueConfig(), valueConfig));
       parent.setLineType(InitCodeLineType.SimpleInitLine);
     }
     return root;
   }
 
+  /**
+   * Parses the path found in {@code scanner} and descend the tree rooted at {@code root}. If
+   * children specified by the path does not exist, they are created.
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java]: s/does/do/

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 16]
```diff
@@ -157,12 +157,13 @@ public static InitCodeNode createSingletonList(String key) {
...
   }
 
+  /** Creates a node to be used as the root of the initialization tree. */
   public static InitCodeNode newRoot() {
     return new InitCodeNode("root", InitCodeLineType.StructureInitLine, InitValueConfig.create());
   }
 
   /*
-   * Constructs a tree of objects to be initialized using the provided context, and returns the root
+   * Constructs a tree of objects to be initialized using the provided context, and returns the root.
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: Make into a one-line java doc (it's not a javadoc now)
`/** ... */`

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 6 above)
Comment (vchudnov-g) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: Urgh.... I don't like mysterious code.

File: src/main/java/com/google/api/codegen/util/Scanner.java [hunk 17]
```diff
@@ -171,7 +171,11 @@ public int scan() {
     return last = codePoint;
   }
 
-  public String token() {
+  public int lastToken() {
+    return last;
+  }
+
+  public String tokenStr() {
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/util/Scanner.java]: We should change the `token` field to `tokenStr` to match, to avoid confusion.

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: I like how you solved this :)

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 18]
```diff
@@ -157,12 +157,39 @@ public static InitCodeNode createSingletonList(String key) {
...
+        if (initValueConfig == null) {
+          root.mergeChild(InitCodeNode.createWithName(nameString, nameString));
+        } else {
+          root.mergeChild(InitCodeNode.createWithValue(nameString, initValueConfig));
+        }
+      }
+    }
+
+    for (String initFieldConfigString : context.initFieldConfigStrings()) {
+      FieldStructureParser.parse(root, initFieldConfigString, context.initValueConfigMap());
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: ping

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 3 above)
Comment (vchudnov-g) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: ok

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 5 above)
Comment (vchudnov-g) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: I agree on not blocking, but we should understand the code.

File: src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java (hunk 2 above)
Comment (vchudnov-g) [reply src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java]: Argh. We should understand the mysterious code, but no need to block this PR on this point

File: src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java (hunk 15 above)
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java]: Done

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java (hunk 16 above)
Comment (pongad) [reply src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: Made into Javadoc, though I think the formatter wants to keep them 3 lined.
